### ISSUE TITLE
Use a single token verifier in the auth middleware

### ISSUE THREAD

**Issue by reactima** [zitadel/zitadel#7937]
## Acceptance criteria

- [ ] Unify the the `VerifySystemToken` with `VerifyAccessToken` and return a single error
- [ ] Resolve the different keysources on query level (yaml  for system user vs DB for regular user)

### Preflight Checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions
- [X] I have joined the [ZITADEL chat](https://zitadel.com/chat)

### Describe your problem

If you are struggling with System API JWT token related issues, here are few tips:

It can be result of:

1. misconfiguring zitadel-tool option  --audience=  --issuer=

https://github.com/zitadel/zitadel-tools

you can double check payload and verify signature with public key here

https://jwt.io/

2. diffrent behavior of openssl and base64 tools on macos and ubuntu

can be caused how/where you generate output of "base64 system-user-1.pub" command 

i didnt dig deeper, but can be caused by stupid thing like trying copy-paste keys into your editor and later differences in encoding or \n \r symbols

3. misconfiguring or typo in username 

```yaml
SystemAPIUsers:
  - system-user-1:
      KeyData: <base64 encoded value of system-user-1.pub>
```      
4. expired token 

To understand all the above I had to go ... very deep into code

I was able to trace my issues by studying APITokenVerifier, VerifySystemToken functions output ... 
https://github.com/zitadel/zitadel/blob/3396657c3f06b1815361de10505b542adb8b9152/internal/api/authz/context.go#L114      

VerifyAccessToken
https://github.com/zitadel/oidc/blob/6d1231cb37c27f29ed6fc52bb52d5e7e260ca2dc/pkg/op/verifier_access_token.go#L31

Recommend to re-read documentation 
https://github.com/zitadel/zitadel/tree/main/docs/docs/guides/integrate/zitadel-apis

And use grpcui
https://github.com/fullstorydev/grpcui

Zitadel code worked as intended, but I have to admit it was a struggle to understand and debug. 

What I was trying to replication with System API is an experience similar to Terraform. 

I didnt want to configure test environment by clicking through all UI features. I wanted to get a POC all configuration options are reproducible through the code.

Hope it helps

### Describe your ideal solution

The ideal solution is just more error handling messages and more documentations

### Version

latest

### Environment

Self-hosted

**Comment by gregkeys** [comment]
I'm struggling with the zitadel-go examples, none of the API examples work and I suspect it's due to a combination of poor documentation and possibly a bug in the code.

Similar to this ticket I traced the error i'm having here
https://github.com/zitadel/oidc/blob/6d1231cb37c27f29ed6fc52bb52d5e7e260ca2dc/pkg/op/verifier_access_token.go#L47

and
https://github.com/zitadel/oidc/blob/6d1231cb37c27f29ed6fc52bb52d5e7e260ca2dc/pkg/oidc/verifier.go#L117

This is the error I am receiving
`2024-07-17 13:35:09 time=2024-07-17T20:35:09.650Z level=INFO msg="oidc introspection" err="ID=OIDC-Eib8e Message=token is not valid or has expired Parent=(issuer does not match: Expected: http://localhost:8080, got: 276407985754931203)"`

the CheckIssuer is throwing the error.

I suspect the issue is between the audience and the issuer being confused somewhere

**Comment by david-driscoll** [comment]
I'm also running into this issue 😢

**Comment by muhlemmer** [comment]
I've edited the main post so we can plan this improvement.

**Comment by yuji-i** [comment]
I  had the same issue.

**Comment by yuji-i** [comment]
I had the same issue. 
I changed both the issuer value and system api user name to the audience value,
then I received the following error:

msg="token verifier repo: verify JWT access token" caller="/home/runner/work/zitadel/zitadel/internal/authz/repository/eventsourcing/eventstore/token_verifier.go:287" error="invalid signature (invalid signature: no possible keys matches)"

I have tested against the latest 2.66.0,

**Comment by yuji-i** [comment]
I changed the jwt issuer, audience, and System API User name  to the URL with port number , something like:
    https://<custom domain>:8080
Still, "issuer does not match" error appeared in the log, though,  it worked !
I finally manage to create a new instance.

**Comment by reactima** [comment]
https://github.com/zitadel/zitadel/blob/d01d003a0326de8ea3102c4b28733322ac9d70cd/internal/api/authz/context.go#L114

VerifyTokenAndCreateCtxData fall back from VerifyAccessToken to VerifySystemToken created lot of confusion for me


https://github.com/zitadel/zitadel/blob/d01d003a0326de8ea3102c4b28733322ac9d70cd/internal/api/authz/context.go#L114

`
func VerifyTokenAndCreateCtxData(ctx context.Context, token, orgID, orgDomain string, t APITokenVerifier) (_ CtxData, err error) {
	ctx, span := tracing.NewSpan(ctx)
	defer func() { span.EndWithError(err) }()
	tokenWOBearer, err := extractBearerToken(token)
	if err != nil {
		return CtxData{}, err
	}
	userID, clientID, agentID, prefLang, resourceOwner, err := t.VerifyAccessToken(ctx, tokenWOBearer)
	var sysMemberships Memberships
	if err != nil && !zerrors.IsUnauthenticated(err) {
		return CtxData{}, err
	}
	if err != nil {
		var sysTokenErr error
		sysMemberships, userID, sysTokenErr = t.VerifySystemToken(ctx, tokenWOBearer, orgID)
		if sysTokenErr != nil || sysMemberships == nil {
			return CtxData{}, zerrors.ThrowUnauthenticated(errors.Join(err, sysTokenErr), "AUTH-7fs1e", "Errors.Token.Invalid")
		}
	}
	var projectID string
	var origins []string
	if clientID != "" {
		projectID, origins, err = t.ProjectIDAndOriginsByClientID(ctx, clientID)
		if err != nil {
			return CtxData{}, zerrors.ThrowPermissionDenied(err, "AUTH-GHpw2", "could not read projectid by clientid")
		}
		// We used to check origins for every token, but service users shouldn't be used publicly (native app / SPA).
		// Therefore, mostly won't send an origin and aren't able to configure them anyway.
		// For the current time we will only check origins for tokens issued to users through apps (code / implicit flow).
		if err := checkOrigin(ctx, origins); err != nil {
			return CtxData{}, err
		}
	}
	if orgID == "" && orgDomain == "" {
		orgID = resourceOwner
	}
	// System API calls don't have a resource owner
	if orgID != "" {
		orgID, err = t.ExistsOrg(ctx, orgID, orgDomain)
		if err != nil {
			return CtxData{}, zerrors.ThrowPermissionDenied(nil, "AUTH-Bs7Ds", "Organisation doesn't exist")
		}
	}
	return CtxData{
		UserID:            userID,
		OrgID:             orgID,
		ProjectID:         projectID,
		AgentID:           agentID,
		PreferredLanguage: prefLang,
		ResourceOwner:     resourceOwner,
		SystemMemberships: sysMemberships,
	}, nil
}

`

**Comment by tafaust** [comment]
Any plans when this bug will be fixed?

**Comment by geekodour** [comment]
what's the resolution here? so is it fair to say instance creation is broken via the systemAPI? but seems like a very well documented process: https://zitadel.com/docs/guides/integrate/zitadel-apis/access-zitadel-system-api

I am not sure why is this broken.

More context:
1.  [SystemAPI access Issue](#orga17fc9c)


<a id="orga17fc9c"></a>

# SystemAPI access Issue

Known/Related issues:

-   <https://github.com/zitadel/zitadel/issues/7937>
-   <https://github.com/zitadel/zitadel/issues/9494>

Following sets up the keys:

    # For generating the private key pair
    system-api-secret:
    	openssl genrsa -out systemapikey/superuser.pem 2048
    	openssl rsa -in systemapikey/superuser.pem -outform PEM -pubout -out systemapikey/superuser.pub
    
    # https://zitadel.com/docs/guides/integrate/zitadel-apis/access-zitadel-system-api
    # generates a token that's valid for 1h
    # export NEW_INSTANCE_ZITADEL_JWT_TOKEN=$(just system-api-access-secret-jwt-key)
    system-api-access-secret-jwt-key:
    	@zitadel-tools key2jwt \
    	--audience=https://auth.duck.com \
    	--key=systemapikey/superuser.pem \
    	--issuer=superuser

Related Zitadel config

    SystemAPIUsers:
      - superuser:
          Path: /systemapikey/superuser.pub

Now we run zitadel, after running `system-api-secret`, so that the public key is being access by our `zitadel start` command. (Is there a way to verify that this is getting picked correctly?)

Then we run:

    export NEW_INSTANCE_ZITADEL_JWT_TOKEN=$(just system-api-access-secret-jwt-key)

This supposedly gives us the key access to our zitadel instance running with our generated public key.

After this we run `create-duck-instance` (a Justfile recepie)

    main_zitadel_instance_url := "auth.duck.com"
    create-duck-instance:
    	MAIN_ZITADEL_INSTANCE_URL={{main_zitadel_instance_url}} \
    	./create_instance.py duck duck auth.duck.com

Following is what `create_instance.py` looks like:

    #!/usr/bin/env python
    # NOTE: This script creates a new virtual instance in ZITADEL.
    # NOTE: Virtual instances are separated at the network level; there is no central UI to view them.
    # See for more details:
    # https://zitadel.com/docs/guides/integrate/zitadel-apis/access-zitadel-system-api
    # https://zitadel.com/docs/apis/resources/system/system-service-create-instance
    
    import http.client
    import json
    import os
    import argparse
    import sys
    
    def create_zitadel_instance(jwt_token, zitadel_url, instance_name, org_name, custom_domain):
        conn = http.client.HTTPSConnection(zitadel_url)
        payload = json.dumps({
          "instanceName": instance_name,
          "firstOrgName": org_name,
          "customDomain": custom_domain,
        })
    
        headers = {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          'Authorization': f"Bearer {jwt_token}"
        }
    
        try:
            conn.request("POST", "/system/v1/instances/_create", payload, headers)
            res = conn.getresponse()
            data = res.read()
            print(f"Status: {res.status} {res.reason}")
            try:
                print(json.dumps(json.loads(data.decode("utf-8")), indent=2))
            except json.JSONDecodeError:
                print(data.decode("utf-8"))
        except Exception as e:
            print(f"An error occurred: {e}", file=sys.stderr)
        finally:
            conn.close()
    
    def main():
        parser = argparse.ArgumentParser(
            description="Create a new ZITADEL instance via the System API.",
            formatter_class=argparse.HelpFormatter
        )
        parser.add_argument("instance_name", help="The name for the new ZITADEL instance (e.g., 'my-company').")
        parser.add_argument("first_org_name", help="The name of the first organization within the new instance.")
        parser.add_argument("custom_domain", help="The custom domain for the instance (e.g., 'auth.my-company.com').")
    
        args = parser.parse_args()
    
        # Get sensitive data from environment variables
        jwt_token = os.environ["NEW_INSTANCE_ZITADEL_JWT_TOKEN"]
        zitadel_instance_url = os.environ["MAIN_ZITADEL_INSTANCE_URL"]
    
        create_zitadel_instance(
            jwt_token,
            zitadel_instance_url,
            args.instance_name,
            args.first_org_name,
            args.custom_domain
        )
    
    if __name__ == "__main__":
        main()

After running this, we get as the API response:

    Status: 401 Unauthorized
    {
      "code": 16,
      "message": "Errors.Token.Invalid (AUTH-7fs1e)",
      "details": [
        {
          "@type": "type.googleapis.com/zitadel.v1.ErrorDetail",
          "id": "AUTH-7fs1e",
          "message": "Errors.Token.Invalid"
        }
      ]
    }

And the following as the logs:

    time="2025-06-18T16:17:25Z" level=warning msg="authz: verify access token" caller="/home/runner/work/zitadel/zitadel/internal/api/authz/context.go:114" error="ID=APP-Reb32 Message=invalid token" org_domain= org_id=
    
    time="2025-06-18T16:17:25Z" level=warning msg="token verifier repo: verify JWT access token" caller="/home/runner/work/zitadel/zitadel/internal/authz/repository/eventsourcing/eventstore/token_verifier.go:287" error="issuer does not match: Expected: https://auth.duck.com, got: superuser"
    
    time="2025-06-18T16:17:25Z" level=warning msg="token verifier repo: decrypt access token" caller="/home/runner/work/zitadel/zitadel/internal/authz/repository/eventsourcing/eventstore/token_verifier.go:283" error="ID=APP-ASdgg Message=invalid token"

**Comment by fforootd** [comment]
The same has come up on [discord](https://discord.com/channels/927474939156643850/1384628280908582923)

I am currently looking into it to understand where the issue originates from

**Comment by geekodour** [comment]
Also it seems like the path is not validated(i.e if i put a non existing path, things still no not report any issues on start, so it's hard to rule this out when error occurs downstream), it's also not clear if it's a relative or absolute path, just jotting this down for later ref.

```
SystemAPIUsers:
  - https://auth.duck.com:
      Path: /somepath.pub
```

**Comment by geekodour** [comment]
Update:

```
Status: 200 OK
{
  "instanceId": "325099177186982763",
  "details": {
    "sequence": "479",
    "creationDate": "2025-06-18T18:10:11.544817Z",
    "resourceOwner": "325099177187048299"
  }
}
```

The issue was filepath! I had an absolute path, in the config! I corrected the absolute path to the actual absolute path.

So the fix here would be to catch the incorrect path mentioned (i.e if no file actually exists, which was my case, we should be raising that error instead of the JWT related errors we're raising)

https://github.com/zitadel/zitadel/blob/28f7218ea1068a4bc7587166ce3a17af20ff4156/internal/api/authz/system_token.go#L89-L98

We're catching the error but think that's getting supressed somewhere?

**Comment by xperrotti** [comment]
just to let you guys know, when you call any API endpoint with a JWT token, Zitadel will try to check if it's a valid system OR human/machine user JWT token. This is why you see invalid token/invalid issuer/invalid signature or whatever. Forget about those, just make sure to:

1. Don't generate your RSA key pair using mac's openssl tool/base64 encoding. It's a pain to get it working, generate a key pair online or on a linux machine and then use the base64 encode website (google it) to encode the public key to base64
2. Make sure to have your system user JWT payload with the 'iss', 'aud' and 'sub' fields set to your zitadel instance's HTTP scheme + FQDN + port as it's logged when you first run the server/start the container/pod. That is, even if you external url is https://zitadel.domain.io, your instance could be internally set to https://zitadel.domain.io:8080 and that breaks the system user jwt check. Check the /ui/console URL that's on the log to check that
3. Don't trust zitadel-tools, at least for now. Use https://dinochiesa.github.io/jwt/ to create the payload and sign the JWT using the key pair created on step 1. Zitadel-tool outputs a JWT with the audience set as a list of strings with 1 value instead of a single string, so I don't trust it

long story short, I think that's it. I've got some thoughts about this in here: https://discord.com/channels/927474939156643850/1277030942606884904

PS: Don't trust any website to create and/or sign the JWT. Your key pair could be leaked, take my advise as an advise and think before you do things

**Comment by fforootd** [comment]
> PS: Don't trust any website to create and/or sign the JWT. Your key pair could be leaked, take my advise as an advise and think before you do things

That is indeed important.

**Comment by BillyBolton** [comment]
@fforootd Any update with this?
